Two slided Instagram flowtiles

elmeraa opened the pull request:
## Summary of changes
- Addition of two targets (`flow_national_instagram_png` and `flow_cartogram_instagram_png`) and functions (`national_ig` and `cartogram_ig`) to plot Instagram sized national flow timeseries (slide 1) and state flow timeseries (slide 2)
- Addition of two output pngs `flow_cartogram_ig.png` & `flow_national_ig.png`. 

Note
- USGS lower left logo is still a bit fuzzy even after increasing width

![flow_national_ig](https://user-images.githubusercontent.com/69014273/208736417-40b447b5-dc14-4010-8ffe-9c1cb0ff5b19.png)
![flow_cartogram_ig](https://user-images.githubusercontent.com/69014273/208736423-aa83db11-e96e-4287-bbb3-caaf69a49b0c.png)

ceenell commented on src/plot_cartogram.R:
```diff
@@ -282,3 +282,253 @@ rm_facet_clip <- function(svg_in, file_out, width){
+               vjust = 1) +
+    # add legend.
+    draw_plot(plot_legend,
+              x = plot_margin*20,
```
Instead of using the `plot_margin` as a multiple, use relative widths to define placement. Use the `plot_margin` to define consistent spacing around the border and for adding padding around the different plot elements.

In this case, the x position ranges from 0 to 1 (left to right). That's defined above in `ggdraw`. In this case your x is 0.5 (`20*0.025`). That's setting where the upper right corner is placed because of the `hjust` and `vjust` settings here. And then you have a negative width.

That seems sort of off from what we actually see in the output though, because I would expect that the legend corner is then in the middle of the plot. There seems to be some unexpected behavior with `plot_legend` and `draw_plot`.

ceenell commented on _targets.R:
```diff
@@ -101,5 +101,27 @@ list(
+  # Flow timeseries nationally - instagram 
+  tar_target(
+    flow_national_instagram_png,
+    national_ig(file_svg = "flow_national_ig.png", 
+                  plot_left = plot_nat, 
+                  date_start,
+                  width = 1080, height = 1080, color_bknd),
```
Since the national chart and state tiles are on separate outputs, it would be smart to rename those targets to be more informative & broadly applicable to export functions

ceenell commented on src/plot_cartogram.R:
```diff
@@ -282,3 +282,253 @@ rm_facet_clip <- function(svg_in, file_out, width){
+              height = 0.37, width = 0.37,
+              hjust = 0, vjust = 1) +
+    # national-level plot
+    draw_plot(plot_left+theme(legend.position = 'none',
+                              axis.title.x.bottom = element_text(size = 14),
+                              axis.title.x.top = element_text(size = 14),
+                              axis.text.x.bottom = element_text(size = 6),
+                              axis.text.y = element_text(size = 8)),
```
Be careful with using many different fonts sizes. Usually I try to stick to just 2

ceenell reviewed (changes requested):
This is looking good. Everything ran for me and I was able to generate the ig pngs. 

Since we now have multiple outputs of the same charts with only a few changes between them, namely dimensions and positioning, this would be a good point to restructure the code a bit. I've added some comments where the code could be a little bit more modular by pulling out repeated steps that are in all the export functions, and moving config-type variables outside of the functions so they can be defined in a single place.

ceenell commented on _targets.R:
```diff
@@ -101,5 +101,27 @@ list(
+  ),
+  
+  # Flow timeseries for states - instagram
+  tar_target(
+    flow_cartogram_instagram_png,
+    cartogram_ig(file_svg = "flow_cartogram_ig.png", 
+                 plot_left = plot_nat,
+                plot_right = plot_cart, 
+                date_start,
+                width = 1080, height = 1080, color_bknd),
+    format = "file"
```
To parallel the steps in the twitter version, you should slightly modify this target and add another after:
```
 # Flow timeseries for states - instagram
  tar_target(
    flow_cartogram_instagram_svg,
    cartogram_ig(file_svg = "flow_cartogram_ig.svg", 
                 plot_left = plot_nat,
                plot_right = plot_cart, 
                date_start,
                width = 1080, height = 1080, color_bknd),
    format = "file"
  ),
  # Remove facet clipping and save as png
  tar_target(
    flow_cartogram_instagram_png,
    rm_facet_clip(svg_in = flow_cartogram_instagram_svg, 
                  file_out = "flow_cartogram_ig.png",
                  width = 16),
    format = "file"
  )
```

The `rm_facet_clip` function is taking the svg output and removing clipping boundaries that are around the drop shadows behind each tile. Without this, there is a strict border that cuts the shadow off abruptly. It's doing this by first saving the tile map as an svg. Then it is hacking into the svg code to delete the `clipPath` elements that are responsible for that.  I'm not sure this will matter much when the tile map is at a much smaller size. I suggest evaluating what it looks like with this extra step or not to decide if it is needed here.

elmeraa replied:
Ahh I see now, I do like them without the strict borders ![flow_cartogram_ig](https://user-images.githubusercontent.com/69014273/209014532-ac32c69d-9521-486f-b417-bdb85e12a938.png)

ceenell replied:
Yeah I agree, the shadow is a little too visually dominant. That could be adjusted in the cartogram plotting function by adjusting [`sigma` or `offset` in `with_shadow`](https://github.com/USGS-VIZLAB/flow-tiles/blob/b5fb678214e12fa1f20aca5a409ad0c0c1cc2980/src/plot_cartogram.R#L52-L59). You'd want to rework the function/pipeline to make that changeable between platforms.

nicolefelts commented:
I can't link it because I'm not logged in to IG on my govt computer, but Elmera, I'd suggest you take a look at the IG version I created of this simply for reference. I had a brief convo with Cee about potential directions for this, and i expect they'll fill you in on some other takeaways from that convo. 
I really like this two slide version you have and I am interested in working together to upload this to our TEST IG account to see how it looks

elmeraa commented on src/plot_cartogram.R:
```diff
@@ -282,3 +282,253 @@ rm_facet_clip <- function(svg_in, file_out, width){
+  plot_month <- lubridate::month(date_start, label = TRUE, abbr = FALSE)
+  plot_year <- lubridate::year(date_start)
+  
+  # import fonts
+  font_legend <- 'Noto Sans Mono'
+  font_add_google(font_legend)
+  showtext_opts(dpi = 300, regular.wt = 200, bold.wt = 700)
+  showtext_auto(enable = TRUE)
```
Wouldn't a simplier approach be `enable_showtext` call in the function then `enable_showtext = TRUE (or FALSE)` in targets?

ceenell replied:
Yeah totally, that would also work. I think you'd want to put it at the beginning and end of `_targets.R` to ensure that it gets run everytime `tar_make()` is run.

elmeraa commented:
## Summary of changes
- added new `restyle_legend` function to then input into `flow_national_instagram_png` & `flow_cartogram_instagram_svg` targets
-  added `flow_label` and `source_label` to top of `_targets.R`
- added `showtext=TRUE` to ensure fonts are enabled
- adjusted `plot_state_cartogram` to have adjustable `sigma`, `x_offset`, and `y_offset` values and added new `plot_cart_ig` target that has adjusted `sigma`, `x_offset`, and `y_offset` values. 
- added `rm_facet_clip` function to `flow_cartogram_instagram_svg` to drop the strict border
- adjusted naming (`plot_left` to `plot_nat` and `plot_right` to plot_cart`) for instagram targets
- added `|> magick::image_scale()` to `usgs_logo` as well as `title_flow` (streamflow title) to avoid fuzzyness of images

## Updates Instagram Versions
![flow_national_ig](https://user-images.githubusercontent.com/69014273/209027428-04d2823d-5378-4873-82cc-97d12eb8484e.png)
![flow_cartogram_ig](https://user-images.githubusercontent.com/69014273/209027434-a3d6e46a-1996-4be4-bc97-b2517f8d3947.png)

ceenell commented on _targets.R:
```diff
@@ -11,8 +11,16 @@ pal_wetdry <- c("#002D5E", "#0C7182", "#6CB7B0", "#A7D2D8", "#E0D796", "#AF9423"
+source_label <- "Data: USGS National Water Information System"
+
+# font - enabeling showtext
+showtext = TRUE
```
Where is this being used in the pipeline?

ceenell commented on _targets.R:
```diff
@@ -11,8 +11,16 @@ pal_wetdry <- c("#002D5E", "#0C7182", "#6CB7B0", "#A7D2D8", "#E0D796", "#AF9423"
+source_label <- "Data: USGS National Water Information System"
+
+# font - enabeling showtext
+showtext = TRUE
```
While this is defined here, it's still being defined in `combine_plots`. We've also got the same sequence of showtext commands being used in multiple places - this snippet:
``` 
font_legend <- 'Noto Sans Mono'
font_add_google(font_legend)
showtext_opts(dpi = 300, regular.wt = 200, bold.wt = 700)
showtext_auto(enable = enable_showtext) 
```
Since `showtext_auto` turns the text on for all functions run after it is enabled, we really only need to run this once per session. We also still need to turn off showtext after we are done, by using `showtext_auto(enable = FALSE)` after plots are drawn. One way to address both of these needs would be to write a wrapper function around the code snippet above, so it is only written in one place. Then this could be run where you define `showtext` here to turn on the fonts, and at the end of the targets list to turn it off. By doing it this way, we don't need to worry about it being run within each function that uses it, because it will get run each session by default.

elmeraa replied:
I'm a bit confused on this. I pulled out the code snipped above to the top of my `_targets.R` script while setting `showtext_auto(enable = TRUE) `  and then at the en
<img width="343" alt="Screen Shot 2022-12-22 at 2 59 35 PM" src="https://user-images.githubusercontent.com/69014273/209239486-cda6dca1-2481-42b2-96aa-4d2798d8a845.png">
d of the pipeline, I added `showtext_auto(enable = FALSE)` but I'm getting this error

ceenell replied:
yes, that's consistent with what I'm suggesting. I have never done this before, so I'm not totally certain what the expected behavior is, but if you didn't make edits to any of the targets then I would expect that output. May require troubleshooting to determine the best way to set this up, and will definitely be valuable to communicate to others as well

ceenell commented on _targets.R:
```diff
@@ -101,5 +110,55 @@ list(
+  # Plot flow timeseries for states with adjusted sigma and offset values - Instagram 
+  tar_target(
+    plot_cart_ig,
+    plot_state_cartogram(state_data = flow_state, fips, pal = pal_wetdry, usa_grid, color_bknd,  
+                         sigma_val = 2.5 , xoffset_val = 0.5, yoffset_val =0.5) 
```
nice, this is a good solution

ceenell commented on src/plot_cartogram.R:
```diff
@@ -101,7 +104,7 @@ plot_national_area <- function(national_data, date_start, date_end, pal, color_b
                        )) +
     theme_flowfacet(base = 12, color_bknd, text_color) +
     theme(axis.text.y = 
-            element_text(size = 12, 
+            element_text(size = 12,
                          vjust = c(1, 0), 
                          hjust = 1),
           axis.title.x.bottom = element_text(size = 20,
```
does it make sense to move these theme elements to `theme_flowfacet`?

elmeraa replied:
This messes with the `plot_state_cartogram` that also uses `theme_flowfacet` ![flow_cartogram](https://user-images.githubusercontent.com/69014273/209222000-4a3731b0-f0b1-4378-a94d-0541a612ae93.png)

ceenell commented on src/plot_cartogram.R:
```diff
@@ -282,3 +286,256 @@ rm_facet_clip <- function(svg_in, file_out, width){
+  showtext_auto(enable = enable_showtext)
+  
+  # Restyle legend
+  plot_nat <- plot_nat +
+    guides(fill = guide_colorsteps(
+      title = "",
+      nrow = 1,
+      direction = 'horizontal',
+      label.position = "bottom",
+      barwidth = 12,
+      barheight = 0.6,
+      background = element_rect(fill = NA),
+      show.limits = TRUE,
+      even.steps = FALSE
+    )) +
+    theme(legend.background = element_rect(fill = NA),
+          text = element_text(family = font_legend, color = text_color, size = 6.5))
+}
```
does it work to return the `get_legend(plot_nat)` here so the function is returning the styled legend rather than the whole plot?

ceenell commented on src/plot_cartogram.R:
```diff
@@ -282,3 +286,256 @@ rm_facet_clip <- function(svg_in, file_out, width){
+               color = text_color,
+               lineheight = 1)  +
+    # stylized streamflow title
+    draw_image(title_flow |> magick::image_scale('800x'),
```
move this `image_scale` up to where the image is first read in

ceenell commented on src/plot_cartogram.R:
```diff
@@ -282,3 +286,256 @@ rm_facet_clip <- function(svg_in, file_out, width){
+              height = 0.37, width = 0.37,
+              hjust = 0, vjust = 1) +
+    # national-level plot
+    draw_plot(restyle_legend+theme(legend.position = 'none',
+                              axis.title.x.bottom = element_text(size = 14),
+                              axis.title.x.top = element_text(size = 14),
+                              axis.text.x.bottom = element_text(size = 6),
+                              axis.text.y = element_text(size = 6)),
```
the naming here is a bit confusing. I would expect that `restyle_legend` is a legend element, but here's it's the plot.   I also think these theme settings can be moved to where the plot or legend are created.

ceenell commented on src/plot_cartogram.R:
```diff
@@ -282,3 +286,256 @@ rm_facet_clip <- function(svg_in, file_out, width){
+              width = plot_margin*60) +
+    # add legend
+    draw_plot(plot_legend,
+              x = plot_margin*20,
```
consider using relative positioning rather than a high multiplier of the `plot_margin` to position this. If `plot_margin` were changed to make more space around the plot, this would move other things around a lot

ceenell reviewed (commented):
A few more comments to address relating to how the legend is styled and created, and separating out the fonts a bit more.

The pngs for the cartogram and national view look good. On the national view, there's 'November' written fairly large, twice. I wonder if it would bemore informative to have the x-axis label be "Day of month" or something similar, and dropping the font size a bit. That way you could increase the size of the chart a little bit more. You could also even the size of the titles a bit so that the month is smaller and streamflow is a little bit bigger.

On the cartogram you may be able to get a little more wiggle room to make the chart bigger if you reduce the strip spacing between the tiles. IDK if this will look good or not though, but could make space to increase the overall tile size. Up to you

ceenell commented:
One comment Nicole had that I think is good - the color legend here is drawn horizontally but the areas in the charts are stacked vertically. it's  a bit counterintuitive to interpreting the charts. Does not need to be address in this PR, but we should think about that more

elmeraa commented:
Chatted with Althea about disabling showtext and theres two approaches:
1) run `c(flowtiles_list, showtext_auto(enable = FALSE))` - this would rerun the pipeline and export pngs with those showtext commands disabled
2) add a note in `README` for users to manually disable showtext commands by running `showtext_auto(enable = FALSE)` in the console

I think option 2 is the best approach here as we don't really want the pipeline to be ran again with the fonts disabled. I've adjusted the `README` to reflect this now. 

@ceenell let me know what you think and if there's any remaining edits - otherwise, I think we can merge in.

ceenell reviewed (approved):
Looks really good @elmeraa !! Can you just add an `out/` directory to export all of the pngs and svgs to before merging?
![flow_cartogram_ig](https://user-images.githubusercontent.com/17803537/209730487-55921f67-6c3f-4151-91d2-0d3121f0ed88.png)
![flow_national_ig](https://user-images.githubusercontent.com/17803537/209730525-66eb603b-8395-4901-9cf8-4069136d67b9.png)
